go-controller/pkg/network-controller-manager/node_network_controller_manager.go

```diff
@@ -0,0 +1,223 @@
+package networkControllerManager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cni"
+	ovncnitypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cni/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/node"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	kexec "k8s.io/utils/exec"
+)
+
+// nodeNetworkControllerManager structure is the object manages all controllers for all networks for ovnkube-node
+type nodeNetworkControllerManager struct {
+	name           string
+	client         clientset.Interface
+	Kube           kube.Interface
+	watchFactory   factory.NodeWatchFactory
+	stopChan       chan struct{}
+	recorder       record.EventRecorder
+	isOvnUpEnabled bool
+
+	defaultNodeNetworkController BaseNetworkController
+
+	// net-attach-def controller handle net-attach-def and create/delete secondary controllers
+	// nil in dpu-host mode
+	nadController *netAttachDefinitionController
+}
+
+// NewNetworkController create secondary node network controllers for the given NetInfo and NetConfInfo
+func (ncm *nodeNetworkControllerManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (NetworkController, error) {
+	topoType := netConfInfo.TopologyType()
+	if topoType == ovntypes.Layer3Topology || topoType == ovntypes.Layer2Topology ||
+		topoType == ovntypes.LocalnetTopology {
+		return node.NewSecondaryNodeNetworkController(ncm.newCommonNetworkControllerInfo(), nInfo, netConfInfo), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topoType)
+}
+
+// newDummyNetworkController creates a dummy network controller used to clean up specific network
+func (ncm *nodeNetworkControllerManager) newDummyNetworkController(topoType, netName string) (NetworkController, error) {
+	netInfo := util.NewNetInfo(&ovncnitypes.NetConf{NetConf: types.NetConf{Name: netName}, Topology: topoType})
+	return node.NewSecondaryNodeNetworkController(ncm.newCommonNetworkControllerInfo(), netInfo, nil), nil
+}
+
+// CleanupDeletedNetworks cleans up all stale entities giving list of all existing secondary network controllers
+func (ncm *nodeNetworkControllerManager) CleanupDeletedNetworks(allControllers []NetworkController) error {
```

Comment:
This cleanup process and `checkForStaleOVSRepresentorInterfaces` have a lot in common and share a lot of logic.

And in this cleanup process, shouldn't we match against running pods as well, just as it is being done in `checkForStaleOVSRepresentorInterfaces`?

So I think we could have:

1. A common utility method that processes the vsctl output providing a map[string]map[string]string that are interfaces to external ids mappings.
2. The cleanup method of the controllers could just call ``checkForStaleOVSRepresentorInterfaces`?

Reply:
cleanup is called when network is deleted. Same as what we discussed before, there should be some other mechanism (Admission webhooks?) to prevent NAD being deleted when Pod still using that NAD. But once cleanup is called, all entities belongs to that network should be deleded, regardless if it is still referenced by a Pod.

On the other hand,  checkForStaleOVSRepresentorInterfaces  is to deleted all the ovs interface that is not referenced by any pod.

Reply:
> cleanup is called when network is deleted. Same as what we discussed before, there should be some other mechanism (Admission webhooks?) to prevent NAD being deleted when Pod still using that NAD. But once cleanup is called, all entities belongs to that network should be deleded, regardless if it is still referenced by a Pod.

So if they are not referenced by a pod, they will be cleaned up in checkForStaleOVSRepresentorInterfaces.

If they are referenced by pod for an unknown network, they will be cleaned up here. But that should not happen as you say, because there should be some other mechanism to prevent NAD being deleted when Pod still using that NAD.

Additionally, even if we clean up here, that does not really free the resource until the actual pod is removed. Or what is actuallly freeing the resource and making it available for use again?

Reply:
here is my understanding: cni should be responsible to delete all the resources that it added in CMDAdd.

The need of checkForStaleOVSRepresentorInterfaces() is only for DPU mode, it also means there are cases that OVS representor OVS interfaces are not successfully deleted when its associated Pod got removed. (honestly I don't know exact case how that happens). That is why we'd also need Cleanup() function to deleted stale OVS interface when the associated network is deleted.

Reply:
> there are cases that OVS representor OVS interfaces are not successfully deleted when its associated Pod got removed. (honestly I don't know exact case how that happens). That is why we'd also need Cleanup() function to deleted stale OVS interface when the associated network is deleted.

Wouldn't representor OVS interfaces not successfully deleted when its associated Pod got removed be cleaned up by `checkForStaleOVSRepresentorInterfaces`?

My first argument here is that we probably don't need this `Cleanup()`.
* representor interfaces not successfully deleted from OVS when its associated Pod got removed will be cleaned up by `checkForStaleOVSRepresentorInterfaces`
* representor interfaces associated to an existing Pod of a network that got removed could be cleaned up. But even if they are, the associated resource won't be freed up until the pod is removed, and until that happens the representor interface can't be consumed for a different pod. So in this case it wouldn't matter if we did not clean it up and just waited for the pod to be removed.

My second argument is that we don't need to do checkForStaleOVSRepresentorInterfaces per network and it could be done globally instead. Since we are removing representor interfaces not associated to any existing pod, it doesn't really matter for which network it was, does it?

Reply:
> Wouldn't representor OVS interfaces not successfully deleted when its associated Pod got removed be cleaned up by checkForStaleOVSRepresentorInterfaces?

it is deleted in delRepPort(), if the ovs interface is already deleted, delRepPort is just a no-op, and will not return failure.

> My first argument here is that we probably don't need this Cleanup().

> representor interfaces not successfully deleted from OVS when its associated Pod got removed will be cleaned up by checkForStaleOVSRepresentorInterfaces
if checkForStaleOVSRepresentorInterfaces() is per-network (see below), then if the network is deleted, no checkForStaleOVSRepresentorInterfaces() would be responsible to delete stale ones for deleted network.

> My second argument is that we don't need to do checkForStaleOVSRepresentorInterfaces per network and it could be done globally instead. Since we are removing representor interfaces not associated to any existing pod, it doesn't really matter for which network it was, does it?

but we'd need to compare the OVS interface iface-id which is network-related to determine it it is the stale representor interface, right?

Reply:
> but we'd need to compare the OVS interface iface-id which is network-related to determine it it is the stale representor interface, right?

Can't we use something else to match? Maybe using sandbox ID or pod UID (`iface-id-ver` external id)?

Reply:
Additionally, making sure I understand the flow for the first argument above. 

Let's say we have a pod running with an attachment to a secondary network via representor:

1. Network is deleted. At this time, we don't remove it's representor from OVS. 
2. Pod is deleted. Since the secondary network controller is not running, it won't remove its representor.
3. At most 1 minute later, checkForStaleOVSRepresentorInterfaces would remove it's representor from OVS.

Even if we removed the representor on step 1, it wouldn't be available for use until after step 2 right?
So the only benefit of removing the representor on step 1 would be to save the minute it could probably take the representor to be available again after step 3.
That benefit might not be very meaningful if the network should not have been deleted before the pod in the first place.

Just trying to understand exactly why would we do it. Not a strong opinion against it.

Reply:
you are already assuming that checkForStaleOVSRepresentorInterfaces () can be done globally, not per-network. I know sandbox external-ids cannot be used as sandbox could change for the same Pod if cni goes through several round of is notBut honestly I am nervous to use sandboxID or iface-id-ver external-ids. In our tests, we already noticed that sandbox could change for the same Pod (the the pod went through several run of cmdAdd/cmdDel) because of any error during cni plumging.

Reply:
Yeah, sorry, I was assuming that without even noticing.

Ok, so sandbox ID is a no-go. We can also potentially add a new external id that would help us in this case (namespace/name?). It seems like it's worth a try but we can discuss further.

Anyway, my original gripe that I probably never ended up mentioning is that there was potential for code reuse between the cleanup and checkForStaleOVSRepresentorInterfaces. The initial vsctl command had only a slight variation. Parsing the output could be common to both methods. And then the vsctl command to remove the interface would also be the same.

I don't want to hold you further on this. So feel free to decide what you want to do here. If on the next meeting we decide to have a global checkForStaleOVSRepresentorInterfaces we can always add it later.